Userland/Libraries/LibJIT/ELFBuild.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2023, Jesús Lapastora <cyber.gsuscode@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibJIT/ELFBuild.h>
+namespace JIT::ELF {
+
+SectionTable::Index StringTable::emit_into_builder(u32 name_index, SectionTable& builder) const noexcept
+{
+    return builder.append(emit_section(name_index));
+}
+
+Section StringTable::emit_section(u32 name_index) const noexcept
+{
+    Elf64_Shdr header {};
+    header.sh_name = name_index;
+    header.sh_type = SHT_STRTAB;
+    header.sh_flags = 0;
+    header.sh_addr = 0;
+    header.sh_link = 0;
+    header.sh_info = 0;
+    header.sh_entsize = 0;
+    header.sh_addralign = 0;
+    return Section(m_data.span(), header);
+}
+
+u64 StringTable::insert(StringView str) noexcept
+{
+    auto const offset = static_cast<u64>(m_data.size());
+    m_data.ensure_capacity(m_data.size() + str.length() + 1);
+
+    for (auto ch : str) {
+        VERIFY(ch != 0);
+        m_data.unchecked_append(ch);
+    }
+    m_data.append(0);
+    return offset;
+}
+
+FixedArray<u8> build_elf_image(u64 shstrndx, ReadonlySpan<Section> sections)
+{
+    Checked<u64> final_image_size = sizeof(Elf64_Ehdr);
+    Vector<u64> section_offsets;
+    section_offsets.ensure_capacity(sections.size());
+
+    auto const sections_begin = final_image_size.value_unchecked();
+    final_image_size += sizeof(Elf64_Shdr) * sections.size();
+
+    for (auto const& section : sections) {
+        auto const offset = final_image_size.value();
+        section_offsets.unchecked_append(offset);
+        if (section.data.has_value()) {
+            final_image_size += section.data.value().size();
+        }
+    }
+
+    auto image = MUST(FixedArray<u8>::create(final_image_size.value()));
```

Comment:
Since this can grow arbitrarily large based on user-provided code, I think this should be TRY instead of MUST.

Reply:
I don't mind changing the APIs to use TRY here, but I don't understand the reasoning. The size of the image is currently constant, it would only differ in the `.strtab`/`.symtab` sizes if we end up adding arbirtary symbol counts. The actual emitted code is not written into the image. Given that we're also assuming that allocations succeed back in `JS::Compiler`, I assume you're thinking about using LibJIT in more memory-tight situations?

Reply:
Oh right. I misunderstood and thought the ELF would contain the code.

Reply:
No, the text section in the ELF image is a NOBITS section pointing to the runtime addresses of the code. Perhaps I should clarify this on a comment or on the commits.